internal/devpkg/package.go

```diff
@@ -556,3 +655,57 @@ func (p *Package) EnsureUninstallableIsInLockfile() error {
 	_, err := p.lockfile.Resolve(p.Raw)
 	return err
 }
+
+// storePath are the constituent parts of
+// /nix/store/<hash>-<name>-<version>
+//
+// This is a helper struct for analyzing the string representation
+type storePathParts struct {
+	hash    string
+	name    string
+	version string
+}
+
+// newStorePathParts splits a Nix store path into its hash, name and version
+// components in the same way that Nix does.
+//
+// See https://nixos.org/manual/nix/stable/language/builtins.html#builtins-parseDrvName
+func newStorePathParts(path string) *storePathParts {
+	path = strings.TrimPrefix(path, "/nix/store/")
+	// path is now <hash>-<name>-<version
+
+	hash, name := path[:32], path[33:]
+	dashIndex := 0
+	for i, r := range name {
+		if dashIndex != 0 && !unicode.IsLetter(r) {
+			return &storePathParts{hash: hash, name: name[:dashIndex], version: name[i:]}
+		}
+		dashIndex = 0
+		if r == '-' {
+			dashIndex = i
+		}
+	}
+	return &storePathParts{hash, name, "" /*version*/}
+}
+
+func (p *storePathParts) Equal(other *storePathParts) bool {
+	return p.hash == other.hash && p.name == other.name && p.version == other.version
+}
+
+// FillNarInfoCache checks the remote binary cache for the narinfo of each
+// package in the list, and caches the result.
+// Callers of IsInBinaryCache must call this function first.
+func FillNarInfoCache(ctx context.Context, packages ...*Package) error {
+	group, _ := errgroup.WithContext(ctx)
+	for _, p := range packages {
+		// If the package's NarInfo status is already known, skip it
+		if _, ok := isNarInfoInCache[p.Raw]; ok {
+			continue
+		}
+		pkg := p
+		group.Go(func() error {
+			return pkg.fillNarInfoCache()
```

Comment:
I don't think `pkg.fillNarInfoCache` is safe to run concurrently. The map, Nix system/version, etc. all need locking.

Reply:
oh good catch! I can ensure the `nix.System` value is cached prior to executing this, but may need to add some locking for the map.